Mathlib/RingTheory/Perfection.lean

```diff
@@ -28,140 +28,216 @@ universe u₁ u₂ u₃ u₄
-indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
-def Ring.perfectionSubring (R : Type u₁) [CommRing R] (p : ℕ) [hp : Fact p.Prime] [CharP R p] :
-    Subring (ℕ → R) :=
-  (Ring.perfectionSubsemiring R p).toSubring fun n => by
-    simp_rw [← frobenius_def, Pi.neg_apply, Pi.one_apply, RingHom.map_neg, RingHom.map_one]
-
-/-- The perfection of a ring `R` with characteristic `p`,
-defined to be the projective limit of `R` using the Frobenius maps `R → R`
-indexed by the natural numbers, implemented as `{f : ℕ → R // ∀ n, f (n + 1) ^ p = f n}`. -/
-def Ring.Perfection (R : Type u₁) [CommSemiring R] (p : ℕ) : Type u₁ :=
-  { f // ∀ n : ℕ, (f : ℕ → R) (n + 1) ^ p = f n }
+/-- The perfection of a monoid `α`, defined to be the projective limit of `α` using the `p`-th
+power maps `α → α` indexed by the natural numbers, implemented as
+`{ f : ℕ → M | ∀ n, f (n + 1) ^ p = f n }`.
 
-namespace Perfection
+If `α` is a ring with characteristic `p` and `p` is prime, `Perfection α p` is also a ring. -/
+def Perfection (α : Type u₁) [Pow α ℕ] (p : ℕ) : Type u₁ :=
+  { f : ℕ → α // ∀ n, f (n + 1) ^ p = f n }
+
+@[deprecated (since := "2025-10-28")] alias Ring.Perfection := Perfection
 
-variable (R : Type u₁) [CommSemiring R] (p : ℕ) [hp : Fact p.Prime] [CharP R p]
+namespace Perfection
 
-instance commSemiring : CommSemiring (Ring.Perfection R p) :=
-  (Ring.perfectionSubsemiring R p).toCommSemiring
+section CommMonoid
 
-instance charP : CharP (Ring.Perfection R p) p :=
-  CharP.subsemiring (ℕ → R) p (Ring.perfectionSubsemiring R p)
+/-- `Perfection M p` as a submonoid of `ℕ → M`. -/
+def submonoid (M : Type*) [CommMonoid M] (p : ℕ) : Submonoid (ℕ → M) where
+  carrier := { f | ∀ n, f (n + 1) ^ p = f n }
+  one_mem' _ := one_pow _
+  mul_mem' hf hg n := (mul_pow _ _ _).trans congr($(hf n) * $(hg n))
 
-instance ring (R : Type u₁) [CommRing R] [CharP R p] : Ring (Ring.Perfection R p) :=
-  (Ring.perfectionSubring R p).toRing
+@[deprecated (since := "2025-10-28")]
+alias _root_.Monoid.perfection := submonoid
 
-instance commRing (R : Type u₁) [CommRing R] [CharP R p] : CommRing (Ring.Perfection R p) :=
-  (Ring.perfectionSubring R p).toCommRing
+instance instCommMonoid (M : Type*) [CommMonoid M] (p : ℕ) : CommMonoid (Perfection M p) :=
+  (submonoid M p).toCommMonoid
 
-instance : Inhabited (Ring.Perfection R p) := ⟨0⟩
+variable (M : Type*) [CommMonoid M] (p : ℕ)
 
 /-- The `n`-th coefficient of an element of the perfection. -/
-def coeff (n : ℕ) : Ring.Perfection R p →+* R where
+def coeffMonoidHom (n : ℕ) : Perfection M p →* M where
   toFun f := f.1 n
   map_one' := rfl
   map_mul' _ _ := rfl
-  map_zero' := rfl
-  map_add' _ _ := rfl
 
-variable {R p}
+/-- The `p`-th root of an element of the perfection. -/
+def pthRootMonoidHom : Perfection M p →* Perfection M p where
+  toFun f := ⟨fun n => coeffMonoidHom M p (n + 1) f, fun _ => f.2 _⟩
+  map_one' := rfl
+  map_mul' _ _ := rfl
```

Comment:
do we have any lemmas about the "spec" of pthRoot in fully applied (unbundled) form like `(pthRoot x) ^ p`, instead of composition?